File: functions/src/api/src/tests/services/identityUploadPayloadContract.test.ts
import { describe, expect, it } from "vitest";
import {
  evaluateNameMatch,
  IDENTITY_UPLOAD_PAYLOAD_CONTRACT_VERSION,
  parseIdentityUploadPayload,
} from "../../services/identityUploadPayloadContract.js";

function buildPayload(header: Record<string, unknown>, body: Buffer): Buffer {
  const headerBytes = Buffer.from(JSON.stringify(header), "utf8");
  const length = Buffer.alloc(4);
  length.writeUInt32BE(headerBytes.length, 0);
  return Buffer.concat([length, headerBytes, body]);
}

describe("identityUploadPayloadContract", () => {
  it("parses v2 contract payload with client information", () => {
    const payload = buildPayload(
      {
        contractVersion: IDENTITY_UPLOAD_PAYLOAD_CONTRACT_VERSION,
        document: {
          documentType: "passport",
          contentType: "image/jpeg",
        },
        clientInfo: {
          firstName: "Ada",
          middleName: "Lovelace",
          lastName: "Byron",
          fullName: "Ada Lovelace Byron",
          email: "ada@example.com",
          dateOfBirth: "1990-01-01",
          countryIso2: "GB",
        },
        extraction: {
          candidateFullName: "Ada Lovelace Byron",
          provider: "on_device_ocr_placeholder_v1",
        },
      },
      Buffer.from([1, 2, 3, 4])
    );

    const parsed = parseIdentityUploadPayload(payload);
    expect(parsed.contractVersion).toBe(IDENTITY_UPLOAD_PAYLOAD_CONTRACT_VERSION);
    expect(parsed.clientInfo.fullName).toBe("Ada Lovelace Byron");
    expect(parsed.extraction.provider).toBe("on_device_ocr_placeholder_v1");
    expect(parsed.documentBytes.byteLength).toBe(4);
  });

  it("rejects malformed payload headers", () => {
    const malformed = Buffer.from([0, 0, 0, 99, 1, 2, 3]);
    expect(() => parseIdentityUploadPayload(malformed)).toThrow(
      "Client payload header is malformed"
    );
  });

  it("evaluates likely name matches", () => {
    const match = evaluateNameMatch("Ada Lovelace Byron", "Ada Byron");
    const mismatch = evaluateNameMatch("Ada Lovelace Byron", "Grace Hopper");
    expect(match.isMatch).toBe(true);
    expect(mismatch.isMatch).toBe(false);
  });
});

File: app/src/test/java/net/metalbrain/paysmart/core/features/identity/uploadhelpers/IdentityUploadPayloadContractTest.kt
package net.metalbrain.paysmart.core.features.identity.uploadhelpers

import net.metalbrain.paysmart.core.features.identity.data.ClientInformation
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.nio.ByteBuffer

/**
 * Unit tests for [IdentityUploadPayloadContract], ensuring that the encoding process correctly
 * packs the header metadata and the raw document bytes into a single byte array.
 *
 * The tests verify:
 * - Structural integrity of the encoded byte array (header length prefixing).
 * - Correct serialization of the JSON header, including contract versioning and document metadata.
 * - Proper inclusion of the raw document payload.
 */
class IdentityUploadPayloadContractTest {

    @Test
    fun `encode packs header and document bytes`() {
        val payloadBytes = byteArrayOf(10, 20, 30, 40)
        val encoded = IdentityUploadPayloadContract.encode(
            IdentityUploadPayloadInput(
                documentType = IdentityDocumentType.PASSPORT,
                contentType = "image/jpeg",
                documentBytes = payloadBytes,
                clientInformation = ClientInformation(
                    firstName = "Ada",
                    middleName = "Lovelace",
                    lastName = "Byron",
                    email = "ada@example.com",
                    dateOfBirth = "1990-01-01",
                    countryIso2 = "gb"
                ),
                extraction = IdentityUploadExtractionSnapshot(
                    candidateFullName = "Ada Lovelace Byron",
                    provider = "on_device_ocr_placeholder_v1"
                )
            )
        )

        val wrapped = ByteBuffer.wrap(encoded)
        val headerLength = wrapped.int
        val headerBytes = ByteArray(headerLength)
        wrapped.get(headerBytes)
        val documentBytes = ByteArray(wrapped.remaining())
        wrapped.get(documentBytes)

        val headerJson = String(headerBytes, Charsets.UTF_8)
        assertTrue(headerJson.contains("\"contractVersion\":\"$IDENTITY_UPLOAD_PAYLOAD_CONTRACT_VERSION\""))
        assertTrue(headerJson.contains("\"documentType\":\"passport\""))
        assertTrue(headerJson.contains("\"countryIso2\":\"GB\""))
        assertTrue(headerJson.contains("\"provider\":\"on_device_ocr_placeholder_v1\""))
        assertEquals(payloadBytes.size, extractByteLength(headerJson))
        assertArrayEquals(payloadBytes, documentBytes)
    }

    private fun extractByteLength(headerJson: String): Int {
        val marker = "\"byteLength\":"
        val index = headerJson.indexOf(marker)
        assertTrue(index >= 0)
        val tail = headerJson.substring(index + marker.length)
        return tail.takeWhile { it.isDigit() }.toInt()
    }
}
